components/blitz/src/ome/services/roi/GeomTool.java

```diff
@@ -354,6 +358,143 @@ public void handle(int x, int y) {
+        // in order to not have to read the same z/t twice
+        final Map<String, List<ome.model.roi.Shape>> zt_lookup =
+            new HashMap<String, List<ome.model.roi.Shape>>();
+
+       // fetch shapes from db and perform some basic checks
+       List results = 
+           session.createQuery(
+               "select distinct s from Shape s " +
+               "left join fetch s.transform t left join fetch s.roi r " +
+               "join fetch r.image i join fetch i.pixels p " +
+               "where s.id in (:ids)").
+           setParameterList("ids", shapeIds).list();
+       if (results.size() != shapeIds.size()) {
+           throw new ApiUsageException("Given shape id(s) invalid");
+       }
+
+       ome.model.core.Image image = null;
+       ome.model.core.Pixels pixels = null;
+       for (final Object r : results) {
+           final ome.model.roi.Shape shape = (ome.model.roi.Shape) r;
+           final ome.model.roi.Roi roi = shape.getRoi();
+           final ome.model.core.Image img = roi.getImage();
+           
+           // check if all shapes come from the same image
+           if (image == null) {
+               image = img;
+               pixels = image.getPrimaryPixels();
+           } else if (!image.getId().equals(img.getId()))
+               throw new ApiUsageException("All shapes have to be from the same image");
+           // check if z/t unattached fallback values are not out of bounds
+           if (zForUnattached < 0 || zForUnattached >= pixels.getSizeZ() ||
+               tForUnattached < 0 || tForUnattached >= pixels.getSizeT())
+               throw new ApiUsageException(
+                   "Fallback value(s) for unattached z/t shapes are out of bounds");
+           // if we have unattached z/t we use the unattached z,t fallback
+           final int theZ = shape.getTheZ() != null ? shape.getTheZ() : zForUnattached;
+           final int theT = shape.getTheT() != null ? shape.getTheT() : tForUnattached;
+           String lookupKey = theZ + "/" + theT;
+           List<ome.model.roi.Shape> lookupValue = zt_lookup.get(lookupKey);
+           if (lookupValue == null) {
+               lookupValue = new ArrayList<ome.model.roi.Shape>();
+               zt_lookup.put(lookupKey, lookupValue);
+           }
+           lookupValue.add(shape);
+       }
+
+       // any point iteration in a tiled image is a lost cause
+       if (data.requiresPixelsPyramid(pixels)) 
+           throw new ApiUsageException("This method can not handle tiled images yet.");
+       // check for channels filter
+       Set<Integer> validChannels = null;
+       if (channels != null && channels.length > 0) {
+           validChannels = new HashSet<Integer>(channels.length);
+           for (int ch : channels) {
+               if (ch < 0 || ch >= pixels.getSizeC())
+                   throw new ApiUsageException("Given channel(s) out of bounds.");
+               validChannels.add(ch);
+           }
```

Comment:
if channels is ``null`` it might be easier to easier to always create ``validChannels``
and iterate through that list
so we can avoid check in 2 locations below i.e.
```
validChannels != null ? validChannels.size() : pixels.getSizeC();
```
and
```
if (validChannels != null &&  !validChannels.contains(c)) continue;
```

Reply:
The first check is a bounds check and targets the given input parameter which is a chance to limit the channels to only a few. The second check is whether the specific channel is in what's that (now bound-checked) list.
I don't see how I can do away with it if I want the caller to restrict the iteration to just a few channels. At a minimum I have to check if it is empty when it iterates over all channels. That's much of a muchness.
Maybe you are saying that the default behavior for an empty input of channels is not ok because in that case all channels are used. If I changed that then we could loop over all channels in the list after bound checking. But that makes the channel input list mandatory from a callee's perspective.

Reply:
I am suggesting to always populate a list of channels index
if int[] channels is not null then we use the specified channels and keep the check
otherwise populate the ``validchannels`` list using value c = 0 to  pixels.getSizeC()-1;

You then iterate through ``validchannels`` for the stats

Reply:
ok, then I read your comment the right way I believe. the very last commit should do that.

Reply:
The comment above depends on what we decide to do for "unattached" channels